### ISSUE TITLE
Meaningful popup characters for the Turkish language pack

### ISSUE THREAD

**Issue by MusKaya** [AnySoftKeyboard/AnySoftKeyboard#3038]
Turkish language pack has tons of popup characters that don't make the slightest sense (this is coming from a native speaker). I'm proposing to get rid of them and instead add symbols that are used frequently (such as exclamation and question marks and so on). I'm willing to create a PR for this if the proposal is accepted.

**Comment by MesutBa** [comment]
Hi,

I have just received the update containing this PR and it came to me as a great annoyance than anything else.

I also am a native Turkish speaker and I live abroad and what has been removed for me was the greatest feature of this keyboard because it allowed me to type in the various languages that I use in daily life (English, French, Turkish) without having to switch keyboards and this is an insanely productive feature.  Let me also add the fact that there are millions of Turkish people 
living abroad.

As of this morning, I find myself unable to type a simple 'é' although this character is already included in the native Turkish Q keyboard layout !  https://kbdlayout.info/KBDTUQ/  (See left of '1' key)

Also, they actually **are** accents in the Turkish language even if they aren't used frequently.  Some prominent examples are: ê, â, î.

Regarding the initial issue by @MusKaya 
> Turkish language pack has tons of popup characters that don't make the slightest sense (this is coming from a native speaker)

Besides the fact that the notion of "meaningful popup characters" is entirely subjective, to me this is about use cases rather than a language issue because we're talking about optional/long-pressed characters here.  In that sense, I'd argue that being a native speaker doesn't add much weight to this and with the tendency to remove anything that appears unuseful, I guess we'll end up with pretty much nothing. There are lots of things that aren't of use to me but I know they're there because someone else makes use of them.

> I'm proposing to get rid of them and instead add symbols that are used frequently (such as exclamation and question marks and so on). 

As the author suggests to "get rid" of these characters, I suppose they came in as an annoyance, however this isn't the case for everyone.  Not only there isn't a single source provided for "symbols that are used frequently", it has been completely dismissed that there are dedicated keys/modes for those, meanwhile there isn't a single other option for all the characters that you removed.

So with this change some characters that do exist become impossible to type and to me that doesn't make the layout more useful, it only cripples it.

Many other keyboard layouts have these optional characters, say on an English layout I have all the options when I press on an 'e' or 'a' key so I fail to understand why all of a sudden I'd stop to get these characters if I'm on a Turkish layout ?

Since they are not being removed from all other layouts because one person found them to be an annoyance, I find this to be a inconsistent too.

Moreover, one of the greatest features of the Turkish Q keyboard is the accent modifier.  Again, you can check the layout https://kbdlayout.info/KBDTUQ/ and look closely to the 'Ş' and ',' keys.  After pressing AltGr + one of the keys mentioned, the following character that your press gets the accent requested.  So I'd go as far as saying that it's in the nature of this keyboard layout to be able to easily type characters outside the Turkish language.

Indeed I do understand that what is offered as part of AnySoftKeyboard may not be an exact replica of the Turkish Q layout.  However, as its name suggests it, compliance is to be expected.

Additionally, I'd also like to note that this type of feature shouldn't be bound to one person's request if the specifications for the keyboard layout aren't clear. In my opinion this development does way more harm than good and I'd like to see it reverted for the standard layout.

If some see a great benefit in using the keyboard in this way, perhaps it can be proposed as an additional "simplified layout" instead of crippling the original one and making it useless.

Sorry for the long long message, and thank you for your time.

**Comment by menny** [comment]
These are excellent points.
Should I revert the change completely and add a new simplified layout from the current? Or something else?

**Comment by MesutBa** [comment]
Thank you for your consideration.
I think a revert + simplified layout would be great.
Thank you so much for your reactivity and for this application ! 🙏🏽❤

**Comment by MusKaya** [comment]
I think reverting is a bad idea. @MesutBa you have a pretty specific use-case:

> for me was the greatest feature of this keyboard because it allowed me to type in the various languages that I use in daily life (English, French, Turkish) without having to switch keyboards

I don't think what we are trying to achieve with a language pack X is the ability to type X plus a variety of other language characters at the same time. Language pack X should offer the best experience for when typing X. Having it also fit the needs of typing languages Y or Z will just degrade the experience for X.

> they actually are accents in the Turkish language even if they aren't used frequently.

This correct. [The only accented characters in that are used in Turkish](https://tr.wikipedia.org/wiki/D%C3%BCzeltme_i%C5%9Fareti#Kullan%C4%B1ld%C4%B1%C4%9F%C4%B1_harfler) are â, ê and î. Since these are not used frequently, it was an oversight to remove them and I agree that we should add it back. **Any other accented character is not a part of Turkish language.** That includes é, even-though it is found on the Turkish Q Keyboard layout.

> this type of feature shouldn't be bound to one person's request

I agree. I also stand by that the current (updated) Turkish layout is far more usable for almost anyone who doesn't need the accented characters that are required by other languages. However I'm also open to add them as popup characters as my first priority with this PR was to replace the default popup characters with frequently used punctuation characters - such as ? ! : )

To summarize my response on this:
1) We need to re-add â, ê and î as they are part of (and the only accented characters) in Turkish language
2) We should not revert the current layout as I think it offers a much better experience for when typing Turkish (even-though this may not be the case for when typing a multitude of languages at the same time)
3) We can add additional accented characters as popup characters and I'm willing to create a PR for this if we have a consensus.

**Comment by menny** [comment]
Can you create a PR for those?

**Comment by MusKaya** [comment]
Yes, but how about receiving a feedback or more on this from @MesutBa and others? Additionally we need to determine the accented characters to be added, @MesutBa can you help compile that list?

**Comment by menny** [comment]
We can/should discuss these changes in a PR where is it easier to see the proposed change

**Comment by MusKaya** [comment]
> We can/should discuss these changes in a PR where is it easier to see the proposed change

@MesutBa can you create a PR with the list of accented characters to add? :)

**Comment by MesutBa** [comment]
### Refusal of the revert addressed

> I think reverting is a bad idea. @MesutBa you have a pretty specific use-case:

@MusKaya  No, actually is the exact opposite, because **what you have done is specific to the point where you are the only person to understand the logic behind it** (if there was any logic at the first place) and you don't even care to explain it.  You are removing existing and valid characters so you must provide a reason to do so because you're changing the defaults.  **You provide no design scheme, no data, no research. Nothing !**  All you say it that it's better and nobody knows why. These are just claims and they aren't backed by any evidence supporting them.

Let me add this simply, physical keyboards are defined by standards because It is a resource used in common by millions of people.  The mobile experience attempts to get as close as possible to it.  So this is not like defining shortcuts on your gaming keyboard because it is better for you.  You cannot change intelligible, common defaults without providing any evidence and claim that it's better while pretending that others have specific use cases.

What I want are optional characters, which already were there, **like on any other keyboard**, I'm not even looking for anything special. I never asked for anything, I never had to.

**Optional** characters are optional as the name tells it. They are useful in the mobile scenario due to the lack of space available on the screen.  They are there to help with as many options as possible because there's no way of doing it like with Alt keys as on a physical keyboard.  So each and every option is in relation to the key pressed as it helps people to know where to look for a character. 

You have removed the ability to type **any** of them. This is certainly **not my personal use case**.  What you have replaced them with though, is definitely _your personal use case_ because the signs you've replaced them with are **irrelevant**.  They do not correlate with the keys you assigned them to.

The signs though, have their own pad with the touch of a dedicated button because it is often not very obvious to assign them to letters.

These were options after all, they were not bothering anyone until you have decided that way and this requires explanation because there are people who use them and you haven't given any valid explanation so far.

So a revert is more likely to be of an absolute necessity rather than a bad idea as you pretend it.

### Points expressed in detail, with examples (summary at the end). I strongly recommend @MusKaya that you read the whole thing though.

> I don't think what we are trying to achieve with a language pack X is the ability to type X plus a variety of other language characters at the same time. Language pack X should offer the best experience for when typing X. Having it also fit the needs of typing languages Y or Z will just degrade the experience for X.

Care to explain **why all well known layouts have tons of optional characters** then ?  **Multi-lingual support is a thing, whether you personally need/like it or not, competing apps support it**.  Is there any rationale behind your claim that basically boils down to say that **someone using the Turkish layout should only type words in the Turkish language ?**  Forget about foreign languages, is it really that hard to think that one may want to type a **proper noun having characters outside the Turkish language** for example ?  I sincerely do not belive that you have put any thought into your implementation for a single second. All you claim since the beginning is that some characters do not have meaning for you.  So be it, but **that's just for you** and no one is forcing you to use them becasue they are optional.  So you have to respect others making use of them.

> This is correct. [The only accented characters in that are used in Turkish are â, ê and î](https://tr.wikipedia.org/wiki/D%C3%BCzeltme_i%C5%9Fareti#Kullan%C4%B1ld%C4%B1%C4%9F%C4%B1_harfler). Since these are not used frequently, it was an omission to remove them and I agree that we should add them back.

@MusKaya You really don't look like you have studied your subject at all, the link you provide is about the **circumflex diacritic** but the list we're after here are ALL the **diacritics**.

The full list for the Turkish language is as follows: `Â`, `â`, **`Ç`**, **`ç`**, **`Ğ`**, **`ğ`**, `Î`, `î`, **`İ`**, **`ı`**, **`Ö`**, **`ö`**, **`Ş`**, **`ş`**, **`Ü`**, **`ü`**, `Û`, `û`.  The ones in bold appear directly on the layout, the others were options until you removed them.

1. So the fact that some of these diacritics appear straight on the layout is because of their frequency of use and this is **definitely** a language issue. That's what makes the Turkish keyboard layout.
2. The ones that do not appear on the default layout **must** be provided as options because what you're trying to achieve is a Turkish keyboard and it needs to provide all the characters necessary to write in the Turkish language.

> **Any other accented character is not a part of Turkish language.** 

I do not know what is your level of certainty while saying this, because it is quite clear that you're not a linguist but let's assume this is true.  So what then ? Is the the `$` or the `₿` sign anymore Turkish ?  Are you really suggesting that someone using one particular layout should be bound to use only and only that language ?  What about **historians, linguists, translators, people working in international businesses such as trade, finances** etc. ?  **Do you actually think they need your approval to be able to use optional characters ?**  I already asked above, why other layouts are providing all these optional characters then ?  What makes you a legitimate persone to remove others' options ?
What you say is profoundly unsound because there is **absolutely nothing** that tells that people using a Turkish layout can only write in the Turkish language. **You have to back your claims with evidence, otherwise they are bogus.**

Besides, **how would you type your messages here, or code and make pull-requests if your Turkish keyboard didn't include `W`, `X` or `Q` which AREN'T in the Turkish alphabet** ? With your reasoning, they shouldn't be there either ?  

> That includes é, even-though it is found on the Turkish Q Keyboard layout.

Have you considered to think for a couple of minutes why the authorities have put it on the **official** keyboard layout at the first place then ?  Same as above, what about `W`, `X` or `Q` then ?  What makes you qualified as a person to remove something that exists on the **official** keyboard ?  Why should we trust you ?  

> I agree. I also stand by that the current (updated) Turkish layout is far more usable for almost anyone who doesn't need the accented characters that are required by other languages. However I'm also open to add them as popup characters as my first priority with this PR was to replace the default popup characters with frequently used punctuation characters - such as ? ! : )

The two punctuation signs you mention were already there to begin with, you didn't add them.  If you added them as options again, then why did you do that ?  What is it that makes your layout _far more usable_ ?  I already had `(`and `)` above the `,` key. Now I have them again, one being atop `Ş` and the other on the `İ`. There already is a dedicated key for `?` and now it appears as an option to `Ç`. Is this what makes **your personal layout** so much better ?  Putting the same signs over and over again as options ?  **You aren't adding anything behind your claims** except _smileys_, care to elborate please ?  Do you have **any data or reseach** behind this ?  **Any frequency analysis, linguistic study, keyboard design rule** ?  Anything really ? Otherwise these are just bold statements but that doesn't make them any meaningful.

> for almost anyone who doesn't need the accented characters that are required by other languages

So people who also type in other languages, or type proper nouns should be banned from using the Turkish layout ?  Honestly, what are you trying to achieve ?  **I can switch to the English layout and write in French, are you seriously pretending that I shouldn't ?**

> We need to re-add â, ê and î as they are part of (and the only accented characters) in Turkish language

No they're not, already answered above and even worse, the link you provided already has more characters than you mention.

> We should not revert the current layout as I think it offers a much better experience for when typing Turkish (even-though this may not be the case for when typing a multitude of languages at the same time)

**Much better according to what/whom ?  Any design rationale, any data, any research, any analysis ?**

Quick return of experience of the current layout: 

* For the last couple of days, I had to switch back and fourth between English and Turkish layouts to use diacritics (Not just for French but also for Turkish because historical texts have them a lot more than contemporary ones).  The number of keys aren't the same, so the shapes too and that makes switching generally annoying if it is done frequently.

* The new optional characters are absolute nonsense to me. It is clear to me that there's no rationale, no data nor research behind it. Let me give you a couple of examples about this supposedly far more usable layout and why it makes me think it is nonsense:

1. The optional characters provided for `n` are `" “ ”`.  Meanwhile, just below `n` at the bottom right we have the `'` key **already having ALL these options and more**, so adding this atop the `n` is only **redundant and has zero added-value** even though it is claimed to be far more superior.
2. The optional characters provided for `f` are `% ‰`.  The word for percentage in Turkish is _yüzde_ so it starts with a `y`. Why would anyone look for a percent sign atop the `f` key while there's a whole optional keyboard button dedicated to all these signs ?  `‰` appears as an option `to` % on the **{}_** numerical keyboard by the way.

I could go on forever like this.  Except the top row (where the options are numbers followed by [] at the top right) all the new options are absolutely useless, there isn't one single reason to find them where they are. They have nothing to do the with the key pressed and that makes them **far less useful** contrary to the bold claim made by the author.

When I press `E` on an English layout, I get `3` followed by a diacritics of `E` and also the Euro sign `€` because **they are related in an intelligible way**, they all start with an `E` and `3` corresponds to the position on the keyboard.

On @MusKaya's pretendedly superior layout when I press `T` I only get `5` and that makes me wonder where to find the Turkish Lira sign. Guess what, the Turkish Lira sign `₺` is to be found on `D` !  What you pretend to be much better is just **optional signs placed the way you felt like it** and no one can get anything meaningful out of it, except you I guess.

Even worse, for the particular case of the `€` sign, it is obtained as an Alt character of the `E` key on a Turkish physical keyboard and for any layout on any mobile phone that I have used during the last decade, it was atop the `E` key.  But @MusKaya **decided** that it has to be on `D` and **we don't know why and he's doesn't explain anything but he says it's better** so we have to live with it ?

### Summary

@MusKaya  Basically, you have removed all optional characters; **including ones belonging to the Turkish language**, from the Turkish keyboard layout and replaced them with **random signs that you claim to be used frequently** without considering usage habits or anything else and you're pretending all these to be "far more useful".  Besides, you have provided zero evidence for why all of that should be better. But you keep insisting that your development shouldn't be reverted because it's better.  I'm having a very difficult time to take all this seriously. **When I first experienced the outcome of your development, I thought it was a bug ! It though it was done inadvertently !  I even sent an email reporting the bug !**  In all honesty, this can only be called the "The Turkish Q layout but with only signs as options because Mustafa Kaya wanted it that way". That's the sole reason of its existence. There's no rationale or thinking behind it. **It's a defect, it's not a feature.**

@menny  There's nothing related to language issues, localisation, ergonomics or user-friendliness in this implementation.  I stand behind my points which the author refused to address time and again, except making some bold claims without providing any evidence in return.  Obviously, if this would have been done as a separate layout no one would complain but it changes the main layout with a considerable amount of removals. It is invasive regarding usage habits and this becomes drastic in terms of ergonomics because we're talking about a keyboard app here.  None of these seem to be taken into account at all. The new defaults do not resemble the slightest to the defaults of any other keyboard app or to the official layout.  Besides, if for a reason that I'm unable to grasp, this were to be the main layout to go with, a careful person would think of providing an additional, more complete layout that contained all the characters that were removed right ?  Say Turkish Q Extended Layout for the example's sake.  Why would you penalise people who use them ? So it also doesn't show any sign of care for other users.  With these in mind, I'm sorry to say that but I sincerely do not believe that the author is after anything useful here. The fact that he removed characters belonging to the official layout and also to the Turkish language already demonstrates pretty clearly how inconsiderate he is.  There wasn't any single careful thought put behind this development I'm afraid, a development that affects the daily lives of many people. There are plenty of things removed and additions aren't intuitive because there's nothing that relates these signs to those keys in an intelligible way and if there were any reason, it is not provided.  Many of the additions are also redundant.  So there's absolutely no added value here. There's nothing that makes the app better with this implementation. It is repeatedly being said that is "it is better" and that's all there is and I'm expected to accept that.  I could even imagine a layout without any single optional character which would be consistent on its own because there's a logic behind it, but what has been done here is simply nonsense.  Moreover, it certainly is presumptous from my part, but the way this is defended makes me think that it was just a quick and easy way of getting a PR accepted on a well known public repo in order to advert it on a CV or something.  I sincerely cannot think of anyone else benefiting from this development in any other way. I actually wonder if he's even using the software.

I think a revert is absolutely necessary because there isn't one single valid reason provided for what is removed and what is introduced. Valid reasons are especially required to remove from the defaults I'd argue, because people had these choices and you're taking it away from them.  Additonally, this new main layout doesn't resemble the slightest to any other well known Turkish Q layout found on other apps or the official keyboard. This is also pretty bad for user acquisition.  Developing a simplified version of the main layout can be interesting for some users I suppose, but this has to be done separately and it needs to be considerate I beleive.

Thank you for your time.

**Comment by MusKaya** [comment]
@MesutBa it doesn't look like this is going to convert into an efficient discussion soon and even though I read your entire post, I don't have time for lengthy debates. I stand by my suggestion. If the majority in this community agrees on the same points, I'm willing to work on it and create a PR.

**Comment by MesutBa** [comment]
@MusKaya Obiously, there cannot be any debate, nor discussion since you keep dismissing the reasons I have outlined already.  There are plenty of them and I have backed them with examples, references; taking time to reflect on the matter.  You haven't done the slightest effort to justify the rationale behind your changes.

Your development was perhaps a good idea from your standpoint, but it actually cripples the experience of others.  As it has already been suggested by the main author @menny, you may keep it in a simplified layout, which I think is great for you.

**Comment by menny** [comment]
Please keep the discussion civil.

I stand by my suggestion: the default layout should be as close as possible to the official, physical keyboard. @MusKaya, please add the missing accents to your layout.

Also, @MusKaya , please add an additional Turkish layout that is similar to the one we had before your change. You can name it "Extended Turkish layout" or something as such. If it is too much of an effort, that's okay, I'll do it when I have more time on my hands.

Thank you for your enthusiasm.

**Comment by MusKaya** [comment]
Thanks @menny, I'll probably have time in a week or two to do this.
